File: frontend/src/main/kotlin/me/devnatan/katan/frontend/socket/socket.kt
package me.devnatan.katan.frontend.socket

import me.devnatan.katan.frontend.DynamicBlock
import me.devnatan.katan.frontend.EmptyBlock
import me.devnatan.katan.frontend.logger.Logger
import me.devnatan.katan.frontend.measureTimeMillis
import org.w3c.dom.WebSocket

class Socket {

    private val logger = Logger("GatewaySocket")
    private lateinit var ws: WebSocket

    var onConnect: EmptyBlock? = null
    var onClose: EmptyBlock? = null
    var onError: EmptyBlock? = null
    var onMessage: DynamicBlock? = null

    var state: SocketState = SocketState.CLOSED
    var connectionTime: Long = -1

    fun connect() {
        connectionTime = measureTimeMillis {
            state = SocketState.CONNECTING
            ws = WebSocket("ws://127.0.0.1:8080")
        }

        ws.onopen = {
            state = SocketState.OPEN
            onConnect?.invoke()
            logger.fine("[CONNECTED] ${connectionTime}ms")
        }

        ws.onmessage = {
            onMessage?.invoke(it.data)
        }

        ws.onclose = {
            state = SocketState.CLOSED
            onClose?.invoke()
            logger.err("[CLOSE]")
        }

        ws.onerror = {
            onError?.invoke()
            logger.warn("[ERROR]")
        }
    }

    fun send(message: String) {
        if (state != SocketState.OPEN)
            throw IllegalStateException("Socket isn't connected")

        ws.send(message)
    }

}

File: frontend/src/socket.ts
import { EventEmitter } from "events";
import Logger from "@/logger";

export default class Socket extends EventEmitter {
  private readonly logger: Logger = new Logger("GatewaySocket");
  private readonly endpoint: string;

  private client!: WebSocket;
  private state: SocketState = SocketState.DISCONNECTED;

  constructor(endpoint: string) {
    super();
    this.endpoint = endpoint;
  }

  connect() {
    if (this.state == SocketState.CONNECTED) return;

    this.client = new WebSocket(this.endpoint);
    this.client.onopen = () => {
      this.emit("connect");
      this.logger.info("[CONNECT]");
      this.state = SocketState.CONNECTED;
    };
    this.client.onclose = () => {
      this.emit("disconnect");
      this.logger.info("[DISCONNECT]");
      this.state = SocketState.DISCONNECTED;
    };
    this.client.onmessage = (data: any) => {
      this.emit("message", JSON.parse(data.data));
    };
  }

  disconnect() {
    if (this.state == SocketState.DISCONNECTED) return;

    this.client.close();
  }

  send(message: any) {
    this.client.send(JSON.stringify(message));
  }
}

enum SocketState {
  CONNECTED,

  DISCONNECTED
}
